lib/experimental-default-theme.json

```diff
@@ -0,0 +1,23 @@
+{
+	"global": {
+		"color": {
+			"primary": "#52accc",
+			"background": "white",
+			"text": "black"
+		}
+	},
+	"features": {
+		"typography": {
+			"dropCap": false
+		}
+	},
+	"blocks": {
```

Comment:
The question here is whether "features"/"config" and "style" should have inner keys "global", "blocks" or the opposite. I don't  feel strongly but  I feel we should be consistent between "style" and "features"

Reply:
From the perspective of the theme maintainer, it might be easier to group by block because you can have tons of them and their names can be long like or hard to remember. Words like "features" or "styles" are short and easy to include. However, I don't have strong preferences. Whatever works for the majority is fine here. We can change it also later and map it in PHP or JavaScript. It should be all transparent from the perspective of `useEditorFeature` consumer.

Reply:
I've just finished the last iteration of https://github.com/WordPress/gutenberg/pull/20290 and went with `styles` as the top-level, but will rework it after lunch to use the block names (+ `globals`) as the top-level keys.

Reply:
oh, before I go down the path, I was thinking we'd have this:

```
{
    "global": {
        "styles": { /* not sure what can go here, nothing or perhaps things we attach to the root container like background color ?*/ },
        "preset": { /* default presets go here */  },
        "features": { /* things enabled by this PR go here */ }
    },
    "core/paragraph": {
        "styles": { /* block styles */ },
        "preset": { /* presets per block, if this is something the block can do */  },
        "features": { /* things enabled per block */ }
    },
    "core/group": {
        "styles": { /* block styles */ },
        "preset": { /* presets per block, if this is something the block can do */  },
        "features": { /* things enabled per block */ }
    }
}
```

Re-reading Riad's message above, you seem to suggest that we can group all blocks under a "blocks" key, right? I think I have a slight preference to avoid that if that's ok.

Reply:
@nosolows, your example is exactly what I was coming to lately 😃

Reply:
That example works for me. I'm not sure whether "features" or "presets" are enough for all "configs" that's why I'm hesitant between having these or a single "config" holding everything.

Put differently, what config will we have:

 - Enable/disable things
 - Provide presets
 - Anything else?

Reply:
Also how do you define the "Global  styles" UI in this config.

Reply:
@nosolosw, can you expand `global` or one of the blocks with a more detailed example? 

Just to make sure we are on the same page, this proposal is targeting `theme.json` that every theme will be able to define.

In the WordPress core, we probably need only the `global` section to provide default values for everything. 

What's the story for core blocks and 3rd party blocks, where would they define `features` or `styles? Does it make sense to allow them in `register_block_type` as part of `settings` object? Can we automatically read them from `theme.json` and inject overrides into existing block definitions on the server and on the client?

Reply:
I guess it'd look like something like this:

```
{
    "global": {
        "features": {
            "dropCap": false,
        },
        "presets": {
            "color": [ /* defines color presets globally */ ],
            "font-size": false, /* disabled globally */
        },
    },
    "core/paragraph": {
        "styles": {
            "color": "red",
            "font-size": 13
        },
        "presets": {
            "font-size": [ /* define font size presets, so they are enabled for this block */ ]
        },
        "features": {
            "dropCap": true
        }
    },
    "core/navigation": {
        "presets": {
            "color": [ /* defines a different color palette */ ]
        }
    }
}
```

It absorbs the things [mentioned here](https://github.com/WordPress/gutenberg/issues/21583#issuecomment-613911005), although I'm unsure about the "global styles ui" bit that @youknowriad mentions. My thinking is: that UI is not different from the block-level inspector controls, part of it should be generated from the implicit block atts of the blocks (line-height controls, etc), the rest is static. Does that sound about right?

However, that makes me think: would we need different configs for different editors (post, site, widgets)?

Reply:
In your example you're disabling the font-size in the "presets". Shouldn't this be done in "features"? Maybe a hint that presets and features are the same thing "config"

Reply:
> My thinking is: that UI is not different from the block-level inspector controls, part of it should be generated from the implicit block atts (line-height controls, etc), the rest is static. Does that sound about right?

In terms of UI, I originally was thinking that UI  should be generated but I changed my mind :) and now more leaning towards the theme saying what things should be editable by UI (palette, a random CSS variable representing a color and used in theme.json in different places..., a random font size CSS variable).

Reply:
Yeah, I did it intentionally to have this kind of discussion. I know that at the current API is different for disable them & configure its values, but I thought that if we use a config file to do this we could benefit from some compression (a single place to do everything).

The only reason I like `presets` at the secondary level is that the data they hold can be long (array of many things with names, slug, etc), unlike the rest of features where most of them would be booleans.

Reply:
At this point of the discussion, I have all my concerned shared and discussed. I'm happy if we move forward with a format and implement things. We didn't decide to remove the "experimental" yet :)

Reply:
> We didn't decide to remove the "experimental" yet :)

I should take that win! 🙈 :tada: 

I feel more comfortable about the direction and format now, we could open it in a release or two. The only bit I'd like to confirm before removing the `experimental` would be whether we want to keep JSON or use PHP as Grzegorz pointed out.

Reply:
too many decision to make :) 

I don't want to make it "non-experimental" too soon for other reasons too. We could also decide to make some changes  to block markup for themes using theme.json (button markup simplification, alignments, color palette hook removal...)

Reply:
> the theme saying what things should be editable by (global styles) UI

So I'm thinking that the whole `theme.json` is tailored to control "the editor", but we haven't defined what's "the editor". I'm not sure we need this, but what you said above sounded to me like we wanted different things for the post editor and the site editor (where global style is). Not sure we should take any decision now, we can perhaps start with what we have and decide later. Thought about raising it anyway.

Reply:
for me theme.json  is about the theme (so the whole  website) and not just the "editor'.

Reply:
I'm not sure if we should have a "global" property. To me, everything we set on theme JSON is global unless we nest it inside a block/template part/slector.

Reply:
@jorgefilipecosta, are you proposing something like:

```
{
    "styles": { /* not sure what can go here, nothing or perhaps things we attach to the root container like background color ?*/ },
    "presets": { /* default presets go here */  },
    "features": { /* things enabled by this PR go here */ }
    "overrides": {
        "core/paragraph": {
            "styles": { /* block styles */ },
            "presets": { /* presets per block, if this is something the block can do */  },
            "features": { /* things enabled per block */ }
        },
        "core/group": {
            "styles": { /* block styles */ },
            "presets": { /* presets per block, if this is something the block can do */  },
            "features": { /* things enabled per block */ }
        }
    }
}
```

This format makes it look more like config for Babel or ESLint :)

Reply:
I actually like more the proposal with global, the config looks simpler both from a human and programmatic perspective (see, [it only requires a loop](https://github.com/WordPress/gutenberg/pull/20290/files?file-filters%5B%5D=.php#diff-4339a9e86722fd029c231f3fb65f0ad2R287)).

Reply:
I'm proposing a shape similar to this one:
```
{
		"styles": {
			"color": "blue",
			"font-size": "14"
		},
		"presets": {
			color:
			gradient:
		},
		"core/paragraph": {
			"styles": {
				"color": "blue",
				"font-size": "14"
			}
		}
}
```

More details at https://github.com/WordPress/gutenberg/pull/20290#pullrequestreview-413787582.

But in summary, the shape of global styles should be equivalent to the shape we have inside a "core/group" or "core/template-part" selectors. If inside the core group/"template part" we don't have a global property the global shape should also not have them.

Reply:
@jorgefilipecosta thanks for sharing. I see one challenge with your proposal. You need to keep the list of all keywords that aren't blocks or template parts to be able to divide what's in the root of the config file.

Reply:
Ah, I see the logic to that proposal, thanks for sharing, Jorge.

My main concern with that proposal would be that I'd like to explore different ways of declaring nested content. I am not too keen on using that approach as it has scale issues with nested levels. I've tried it and it becomes cumbersome to declare only 2 nested levels, 3 are quite unwieldy. In my past experiments with a simple pattern like the "numbered features" I needed 3 levels to be able to encapsulate the pattern, and I presume that's going to be common in some situations if people use the columns block. However! What I learned the past week is that nested content is a huge topic and a discussion for later, so I don't want to make many assumptions just yet.

Note also a big disclaimer here: we can iterate on the shape of theme.json. I'm eager to start with something so we can gather feedback from theme-authors asap on the parts that are missing.

Reply:
> You need to keep the list of all keywords that aren't blocks or template parts to be able to divide what's in the root of the config file.

We already have the list it is "styles" and "presets" everything else is blocks. Keeping a list is not hard as if we added another thing different from styles or presets we need to process the thing so we can easily add to the list.

If we really want to avoid the list another option would be to make the "blocks" inside a blocks key: 
```
{
		"styles": {
			"color": "blue",
			"font-size": "14"
		},
		"presets": {
			color:
			gradient:
		},
		"blocks": {
			"core/paragraph": {
				"styles": {
					"color": "blue",
					"font-size": "14"
				}
			}
		}
}
```

I would prefer a "blocks" key than a global key, as it would keep the shape on nesting.

Reply:
Jorge and I talked about the shape of `theme.json`, and we've agreed to [land what we have](https://github.com/WordPress/gutenberg/pull/20290) (top-level blocks + global) and iterate. How we capture nested content is going to influence the shape of this, but the fact that it's experimental give us leeway to modify it without concerns about backward-compatibility. Probably it's a good idea to not remove the experimental until we figure out that part.

Reply:
Fantastic, thank you for letting me know. I will update the default config to follow that (I still need to rebase). It looks like only need to move `features` to `global` section and leave the rest as is 👍

Reply:
Now that we have https://github.com/WordPress/gutenberg/pull/22518, I updated this PR to follow the recommendations.